sdks/python/apache_beam/runners/common.py

```diff
@@ -51,6 +51,289 @@ def receive(self, windowed_value):
+  """Represents a method of a DoFn object."""
+
+  def __init__(self, method_value, args, defaults):
+    """
+    Initiates a ``DoFnMethodWrapper``.
+
+    Args:
+      method_value: Python method represented by this object.
+      args: a list that gives the arguments of the given method.
+      defaults: a list that gives the default values of arguments mentioned
+        above. If len(defaults) > len(args) default values are for the last
+        len(defaults) arguments in args.
+    """
+    self.args = args
+    self.defaults = defaults
+    self.method_value = method_value
+
+  def call(self, input_args, input_kwargs):
+    """Invokes the method represented by this object.
+
+    Args:
+      input_args: a list of arguments to be passed when invoking the method.
+      input_kwargs: a dictionary of keyword arguments to be passed when invoking
+                    the method.
+
+    Using a regular method call here instead of __call__ to reduce per-element
+    overhead.
+    """
+    return self.method_value(*input_args, **input_kwargs)
+
+
+class DoFnSignature(object):
+  """Represents the signature of a given ``DoFn`` object.
+
+  Signature of a ``DoFn`` provides a view of the properties of a given ``DoFn``.
+  Among other things, this will give an extensible way for for (1) accessing the
+  structure of the ``DoFn`` including methods and method parameters
+  (2) identifying features that a given ``DoFn`` support, for example, whether
+  a given ``DoFn`` is a Splittable ``DoFn`` (
+  https://s.apache.org/splittable-do-fn) (3) validating a ``DoFn`` based on the
+  feature set offered by it.
+  """
+
+  def __init__(self, do_fn):
+    # We add a property here for all methods defined by Beam DoFn features.
+    self.process_method = None
+    self.start_bundle_method = None
+    self.finish_bundle_method = None
+
+    assert isinstance(do_fn, core.DoFn)
+    self.do_fn = do_fn
+
+    def _create_do_fn_method(do_fn, method_name):
+      arguments, _, _, defaults = do_fn.get_function_arguments(method_name)
+      defaults = defaults if defaults else []
+      method_value = getattr(do_fn, method_name)
+      return DoFnMethodWrapper(method_value, arguments, defaults)
+
+    self.process_method = _create_do_fn_method(do_fn, 'process')
+    self.start_bundle_method = _create_do_fn_method(do_fn, 'start_bundle')
+    self.finish_bundle_method = _create_do_fn_method(do_fn, 'finish_bundle')
+    self._validate()
+
+  def _validate(self):
+    # start_bundle and finish_bundle methods should only have ContextParam as a
+    # default argument.
+    self._validate_start_bundle()
+    self._validate_finish_bundle()
+    self._validate_process()
+
+  def _validate_start_bundle(self):
+    self._validate_bundle_method()
+
+  def _validate_finish_bundle(self):
+    self._validate_bundle_method()
+
+  def _validate_process(self):
+    pass
+
+  def _validate_bundle_method(self):
+    assert core.DoFn.ElementParam not in self.start_bundle_method.defaults
```

Comment:
I wonder if this could be a positive assertion instead, so any *ParamType that's added wouldn't need to get listed here.